pmd-java/src/test/java/net/sourceforge/pmd/lang/java/oom/DataStructureTest.java

```diff
@@ -30,13 +46,39 @@
  */
 public class DataStructureTest extends ParserTst {
 
+    MetricKey<ClassMetric> classMetricKey = new MetricKey<ClassMetric>() {
+        @Override
+        public String name() {
+            return null;
+        }
+
+
+        @Override
+        public ClassMetric getCalculator() {
+            return new RandomMetric();
```

Comment:
Seems like idiomatic support would be nice for these things, so assuming Java 8:
```
MetricKey<ClassMetric> classMetricKey = MetricKey.of(null, RandomMetric::new)
```

With a MetricKey static method:
```
static <T extends Metric> MetricKey<T> MetricKey.of(String name, Supplier<T> supplier)
```

Also, curious, are MetricKey intended to use identity based equality?  The JavaDoc doesn't say, seems useful for a "key" object.  The non-test implementations I see are `enum` based.